paddle/fluid/framework/framework.proto

```diff
@@ -91,52 +91,54 @@ message OpProto {
   required string comment = 5;
 }
 
-enum DataType {
-  BOOL = 0;
-  INT16 = 1;
-  INT32 = 2;
-  INT64 = 3;
-  FP16 = 4;
-  FP32 = 5;
-  FP64 = 6;
-}
+message VarType {
+  enum Type {
+    // POD types
+    BOOL = 0;
+    INT16 = 1;
+    INT32 = 2;
+    INT64 = 3;
+    FP16 = 4;
+    FP32 = 5;
+    FP64 = 6;
+
+    // other types, that might need additonal descriptions
+    LOD_TENSOR = 7;    // LoDTensor
+    SELECTED_ROWS = 8; // Tensor
+    FEED_MINIBATCH = 9;
+    FETCH_LIST = 10;
+    STEP_SCOPES = 11;
+    LOD_RANK_TABLE = 12;
+    // LOD_TENSOR_ARRAY = 7; // replaced by array-of-tensors. c.f. ArrayType
+    // PLACE_LIST = 8; // replaced by array-of-place. c.f. ArrayType
+    READER = 13; // ReaderType
+  }
 
-message TensorDesc {
-  required DataType data_type = 1;
-  repeated int64 dims = 2; // [UNK, 640, 480] is saved as [-1, 640, 480]
-}
+  required Type type = 1;
 
-message LoDTensorDesc {
-  required TensorDesc tensor = 1;
-  optional int32 lod_level = 2 [ default = 0 ];
-}
+  message Tensor {
+    required Type data_type = 1;
+    repeated int64 dims = 2; // [UNK, 640, 480] is saved as [-1, 640, 480]
+  }
+  optional Tensor selected_rows = 2;
 
-message LoDTensorArrayDesc {
-  required TensorDesc tensor = 1;
-  optional int32 lod_level = 2 [ default = 0 ];
-}
+  message LoDTensor {
+    required Tensor tensor = 1;
+    optional int32 lod_level = 2 [ default = 0 ];
+  }
+  optional LoDTensor lod_tensor = 3;
 
-message ReaderDesc { repeated LoDTensorDesc lod_tensor = 1; }
+  message Array { VarType elem_type = 1; }
```

Comment:
Should the VarType be repeated?
And `message Array` can be used as `array of VarType` in which the VarType should be the same and `n-tuple` in which the VarType of the array can be different, right?

Reply:
I don't think the `VarType` in the array should be repeated. This is because the array will have all elements of the same `VarType`.  In an n-tuple, the `VarType` will be repeated.

Reply:
I agree with Abhinav's comment.

Reply:
@abhinavarora Thanks, you are right, I finally know where I was wrong.
We should add `message Tuple {repeated VarType tuple_type = 1;}`

Reply:
Yes this has also been discussed here: https://github.com/PaddlePaddle/Paddle/issues/8336 Instead of `VarDesc` we will have `VarType` as has been done in this PR.

Reply:
@abhinavarora mentioned before, can the elements of Fluid's channel be a channel? I have thought, I think it should be supported.

Reply:
Yes, channel of channels is a commonly used pattern used with CSP

Reply:
Maybe I don't understand your question, but from what I understand -- The `ChannelDesc` should have a `VarType` which can be either be a tuple or any of the primitive types supported, a `capacity` and maybe a `name` .

Reply:
If a channel can hold so many types, we cannot make shape inference easily at compile time.

Reply:
I think the `Tuple` and `Array` should be two different types. 

* The `Array` can hold elements with the same type and shape at compile type.  The shape of `array_read` can be inferenced at compile time only if the `array` holds the same time.
  * The length of Array can only be decided at runtime since there are `array_write` operators for RNN. If we make array can hold different types, we cannot inference type at compile time.


* The `Tuple` is an immutable type. The length of `Tuple` and its elements will not be changed at compile time.

Reply:
@reyoung You are right, `Tuple` and `Array` are 2 different types. The type of the array elements will be known at compile time. `Tuple` should be immutable and its length and elements will not change. I am just a little curious how we could implement thin our cpp code. Today, we map the `LODTensorArray` directly to a `Vector<LODTensor>`. For an `Array` would we have to create mappings for all types?

Reply:
> would we have to create mappings for all types?

I am also curious about this.